File: scripts/test_solutions/017_letter_combinations.js
var letterCombinations = function(digits) {
    if (!digits.length) return [];
    const map = { '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz' };
    const result = [];
    function backtrack(idx, path) {
        if (idx === digits.length) {
            result.push(path);
            return;
        }
        for (const char of map[digits[idx]]) {
            backtrack(idx + 1, path + char);
        }
    }
    backtrack(0, '');
    return result;
};

File: scripts/test_solutions/017_letter_combinations.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};

        unordered_map<char, string> mapping = {
            {'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"},
            {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}
        };

        vector<string> result;

        function<void(int, string)> backtrack = [&](int idx, string path) {
            if (idx == digits.size()) {
                result.push_back(path);
                return;
            }
            for (char c : mapping[digits[idx]]) {
                backtrack(idx + 1, path + c);
            }
        };

        backtrack(0, "");
        return result;
    }
};
